### PR TITLE
Support multimodule pipelining in 1F1B schedule

### PR DESCRIPTION
**Pull request by yashaswikarnati**
## Summary

  Adds support for multi-module pipeline parallelism (encoder + LLM) in the 1F1B schedule.

  **Changes:**
  - Add `MultiModuleProcessGroupCollection` for managing process groups across modules
  - Support dict-based tensor format `{module_name: tensor}` in forward/backward
  - Handle 2D/3D tensor conversion for P2P and bridge communication
  - Add `backward_step_multimodule` to handle backward for multimodule cases

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/pipeline_parallel/bridge_communicator.py [hunk 1]
```diff
@@ -494,7 +494,7 @@ def recv_backward(self) -> torch.Tensor:
                 received_gradients_list.append(grad_tensor)
 
             # Concatenate received gradients
-            aggregated_gradient = torch.cat(received_gradients_list, dim=0)
+            aggregated_gradient = torch.cat(received_gradients_list, dim=self.dim_mapping['b'])
```
Comment (yaoyu-33) [inline megatron/core/pipeline_parallel/bridge_communicator.py:517]: can you remind me what's dim_mapping here, and what's "b"

Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/bridge_communicator.py:517]: ah dim mapping is which dimension index correspond to batch, seq and hidden. 'b' is for batch

File: megatron/core/pipeline_parallel/multimodule_communicator.py [hunk 2]
```diff
@@ -45,6 +45,52 @@ class RankModuleInfo:
...
+    P2P and bridge communicators expect 3D tensors.
+    Handles both single tensors and lists of tensors (for VPP).
+
+    Args:
+        tensor: Input tensor (2D or 3D), list of tensors, or None.
+
+    Returns:
+        3D tensor (with singleton last dim if input was 2D), list of 3D tensors, or None.
+    """
+    if tensor is None:
```
Comment (yaoyu-33) [inline megatron/core/pipeline_parallel/multimodule_communicator.py:66]: can you assert fail if a 3d tensor passed in and its last_dim.size != 1

File: megatron/core/pipeline_parallel/schedules.py [hunk 3]
```diff
@@ -2164,27 +2205,27 @@ def enable_grad_sync():
...
         )
-        p2p_communicator.send_forward(output_tensor, is_pp_last_stage(p2p_communicator.pp_group))
+        p2p_communicator.send_forward(output_tensor, p2p_communicator.is_pp_last_stage)
         total_num_tokens += num_tokens
 
         if not forward_only:
             input_tensors.append(input_tensor)
             output_tensors.append(output_tensor)
-            deallocate_output_tensor(output_tensor[0], config.deallocate_pipeline_outputs)
+            deallocate_output_tensor(output_tensor, config.deallocate_pipeline_outputs)
```
Comment (yaoyu-33) [inline megatron/core/pipeline_parallel/schedules.py:2230]: why [0] is removed here?

Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/schedules.py:2230]: to consistently handle individual tensors or lists or dicts we have these checks in the deallocate output tensor. just passing the first element is indeed not needed

File: megatron/core/pipeline_parallel/multimodule_communicator.py (hunk 2 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/multimodule_communicator.py:66]: I think for 3D tensor its a no op. any 3d tensor is fine (last dim can be 1 and no assert is needed). will make it clear in the doc string. thx

**Comment by shifangx** [comment]
Hi, @yaoyu-33, @yashaswikarnati, Could you help increase the priority of this PR?
As for as I know, [#3129](https://github.com/NVIDIA/Megatron-LM/pull/3129) is the last functionality PR for M4, and it is fundamental for DistTrain.

There are also other leftover prs for M4, but [#3129](https://github.com/NVIDIA/Megatron-LM/pull/3129) is the most importance one.

File: megatron/core/pipeline_parallel/multimodule_communicator.py (hunk 2 above)
Comment (yaoyu-33) [reply megatron/core/pipeline_parallel/multimodule_communicator.py:66]: if original 3d and last dim==1, _restore_tensor_from_comm will make it a 2d?
dim in = [a, b, 1], dim out will be [a, b]?

File: tests/unit_tests/pipeline_parallel/test_multimodule_schedules.py [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+    block = DistributedDataParallel(
+        config=block.config, ddp_config=ddp_config, module=block, pg_collection=pg_collection
+    )
+    block.pre_process = False
+    block.post_process = False
+    block.share_embeddings_and_output_weights = False
+    return block
+
+
+def create_module_with_grid(tp, pp, dp, grid_offset, hidden_size):
```
Comment (shifangx) [inline tests/unit_tests/pipeline_parallel/test_multimodule_schedules.py:156]: Maybe `create_module_and_grid` is more appropriate, because this function create grid, not use grid to create model.

**Comment by yashaswikarnati** [comment]
/ok to test [5f941d1](https://github.com/NVIDIA/Megatron-LM/pull/3129/commits/5f941d15fd5545e600ebf4bdba3c568bcbb3cbf7)

**Comment by shifangx** [comment]
> Hi, @yaoyu-33, @yashaswikarnati, Could you help increase the priority of this PR? As for as I know, [#3129](https://github.com/NVIDIA/Megatron-LM/pull/3129) is the last functionality PR for M4, and it is fundamental for DistTrain.
> 
> There are also other leftover prs for M4, but [#3129](https://github.com/NVIDIA/Megatron-LM/pull/3129) is the most importance one.

Hi, @yashaswikarnati , what is the next step of this pr?
 @NVIDIA/core-adlr , @NVIDIA/pipeline-parallelism , @NVIDIA/mcore-oncall , @erhoo82 , can you help review this pr?

**Comment by shifangx** [comment]
Hi, @yashaswikarnati, can you help to address CI test issue?
Some test cases failed.

File: megatron/core/pipeline_parallel/utils.py [hunk 5]
```diff
@@ -355,3 +355,88 @@ def get_comm_stream():
...
+        output_tensor = {language_model_module_name: output_tensor}
+
+    # Handle output_tensor_grad: None (last stage) or dict (intermediate stages)
+    if not output_tensor_grad:
+        # Last stage: no gradient from next stage
+        output_tensor_grad = {key: None for key in output_tensor.keys()}
+
+    # Apply grad scaling if needed (for last stage only)
+    for module_name in output_tensor.keys():
+        if output_tensor_grad[module_name] is None and config.grad_scale_func is not None:
```
Comment (janEbert) [inline megatron/core/pipeline_parallel/utils.py]: Just out of curiosity, why is the scaling only applied when the gradient is `None`?

Comment (janEbert) [reply megatron/core/pipeline_parallel/utils.py]: Doesn't this break when using gradient accumulation?

**Comment by yashaswikarnati** [comment]
/ok to test [ee189df](https://github.com/NVIDIA/Megatron-LM/pull/3129/commits/ee189df4eca79a874bb67c4629bb9a2d85613487)

File: megatron/core/pipeline_parallel/multimodule_communicator.py (hunk 2 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/multimodule_communicator.py:66]: good call out! i added an assert in the prepare comm to prevent this [ee189df](https://github.com/NVIDIA/Megatron-LM/pull/3129/commits/ee189df4eca79a874bb67c4629bb9a2d85613487)

**Comment by yashaswikarnati** [comment]
> Hi, @yashaswikarnati, can you help to address CI test issue? Some test cases failed.

test failures look unrelated and all tests pass locally, retriggered again

File: megatron/core/pipeline_parallel/utils.py (hunk 5 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/utils.py]: This is the same pattern as the existing backward_step in https://github.com/NVIDIA/Megatron-LM/blob/65217aa2b62494d82b0e61378032f1f35020aa43/megatron/core/pipeline_parallel/schedules.py#L477-L478

  output_tensor_grad is None only on the last pipeline stage (no downstream stage to send back a gradient). At the last stage the output is the scalar loss itself, so we scale it before calling backward() with
  grad_tensors=None (PyTorch uses implicit 1.0 for scalars). Intermediate stages already receive scaled gradients via P2P, so scaling again would double-scale.

  Gradient accumulation is fine — this runs per-microbatch, and each microbatch's loss is independently scaled before its backward pass.

Comment (janEbert) [reply megatron/core/pipeline_parallel/utils.py]: That makes so much sense. Thank you for the explanation!

**Review by janEbert — APPROVED** [review]

**Comment by shifangx** [comment]
Hi, @NVIDIA/core-adlr, @NVIDIA/pipeline-parallelism , @erhoo82 , could you please help to review this PR?
This pr is very useful for multimodality training.

**Review by shifangx — APPROVED** [review]

File: megatron/core/pipeline_parallel/schedules.py [hunk 6]
```diff
@@ -2028,43 +2067,45 @@ def forward_backward_pipelining_without_interleaving(
...
-        tp_group = pg_collection.tp
-        cp_group = pg_collection.cp
+        # Custom process groups provided
+
+        if is_multimodule:
+            # Multi-module: use language model's CP size for loss scaling
+            if not config.variable_seq_lengths:
+                raise ValueError(
+                    "config.variable_seq_lengths=True required for multi-module pipelines"
+                )
```
Comment (lhb8125) [inline megatron/core/pipeline_parallel/schedules.py:2084]: Could you help me understand the relationship between ```variable_seq_length``` and ```multi-module pipeline```?

File: megatron/core/pipeline_parallel/utils.py [hunk 7]
```diff
@@ -355,3 +355,88 @@ def get_comm_stream():
...
     return _COMM_STREAM
+
+
+def backward_step_multimodule(
+    input_tensor: Dict[str, torch.Tensor],
+    output_tensor: Union[torch.Tensor, Dict[str, torch.Tensor]],
+    output_tensor_grad: Optional[Dict[str, torch.Tensor]],
+    config,
+    language_model_module_name: str,
+) -> Dict[str, torch.Tensor]:
```
Comment (lhb8125) [inline megatron/core/pipeline_parallel/utils.py]: Maybe move this function to ```schedules.py```?

Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/utils.py]: I can definitely move it there, but was wondering if the schedules.py is already too long to put this in there. also this is pretty self contained

File: megatron/core/pipeline_parallel/schedules.py (hunk 6 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/schedules.py:2084]: the current static shape codepath assume there is only one constant sequence length - which is the language model seq length. in here the PP stages for encoder and LLM can have different seq length and this enforces that shape exchange happens at run time

File: megatron/core/pipeline_parallel/utils.py (hunk 7 above)
Comment (lhb8125) [reply megatron/core/pipeline_parallel/utils.py]: I agree that the ```schedules.py``` is too large, but putting one of the functions of backward_step() in a separate file named ```utils.py``` is a little weird. We can refactor the ```schedules.py``` in another PR. I'd like to hear advice from @erhoo82 @deepakn94

Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/utils.py]: I dont have any strong preference, happy to move it to schedules.py as well. please lmk

File: megatron/core/pipeline_parallel/schedules.py [hunk 8]
```diff
@@ -2028,43 +2067,45 @@ def forward_backward_pipelining_without_interleaving(
...
+            if pg_collection.has_language_model():
+                cp_size = pg_collection.get_language_model_cp_size()
+            else:
+                # Encoder-only ranks: cp_size not used for loss scaling, default to 1
+                cp_size = 1
```
Comment (jaredcasper) [inline megatron/core/pipeline_parallel/schedules.py]: This seems dangerous. What if an encoder ends up using cp? If it isn't used at all maybe setting to None would be safer?

File: megatron/core/pipeline_parallel/schedules.py [hunk 9]
```diff
@@ -2028,43 +2067,45 @@ def forward_backward_pipelining_without_interleaving(
...
     elif p2p_communicator is not None and pg_collection is not None:
-        model_type = get_model_type(model)
-        assert model_type != ModelType.encoder_and_decoder, (
-            "encoder PP stages not yet supported when passing custom process groups. "
-            "support coming soon!"
-        )
-        assert hasattr(p2p_communicator, 'config'), "p2p_communicator must have a config"
-        assert hasattr(pg_collection, 'tp'), "pg_collection must have tp_group"
-        assert hasattr(pg_collection, 'cp'), "pg_collection must have cp_group"
-        assert hasattr(pg_collection, 'embd'), (
```
Comment (jaredcasper) [inline megatron/core/pipeline_parallel/schedules.py:2010]: Why are these asserts removed?

File: megatron/core/pipeline_parallel/utils.py (hunk 7 above)
Comment (jaredcasper) [reply megatron/core/pipeline_parallel/utils.py]: I agree this logically fits with the other backward_step functions in schedules.py, it's not really a util.

File: megatron/core/pipeline_parallel/p2p_communication.py [hunk 10]
```diff
@@ -162,6 +163,21 @@ def __init__(self, pp_group: dist.ProcessGroup, config: ModelParallelConfig):
...
+        """Return True if pp first stage."""
+        return is_pp_first_stage(self.pp_group)
+
+    @property
+    def is_pp_last_stage(self) -> bool:
+        """Return True if pp last stage."""
+        return is_pp_last_stage(self.pp_group)
+
+    @property
+    def num_warmup_microbatches(self) -> int:
```
Comment (jaredcasper) [inline megatron/core/pipeline_parallel/p2p_communication.py]: I don't think this is a "property" of the process group, and it's only calculated once in schedules.py that I see, why move the calculation here?

File: megatron/core/pipeline_parallel/schedules.py (hunk 9 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/schedules.py:2010]: so these process groups are not at all used in the schedules here, these are actually used in finalize_model_grads.py here and the same set of asserts already exist 
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/finalize_model_grads.py#L413-L431

I was wondering that these are redundant here and might have added by mistake before. I agree that this is not the scope of this PR, i can revert as well. please lmk

File: megatron/core/pipeline_parallel/p2p_communication.py (hunk 10 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/p2p_communication.py]: So we made this is a "property" of the communicator object as the calculation differs based on P2PCommunicator vs multimodule case in MultimoduleCommunicator in here 
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/pipeline_parallel/multimodule_communicator.py#L163

File: megatron/core/pipeline_parallel/schedules.py (hunk 8 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/schedules.py]: good callout, for the encoder paths made cp_size to be None

File: megatron/core/pipeline_parallel/utils.py (hunk 7 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/utils.py]: moved to schedules.py

**Comment by shifangx** [comment]
Hi, @erhoo82 , can you help to review this pr?

**Review by lhb8125 — APPROVED** [review]
The code changes of pipeline parallelism look good to me.

File: megatron/core/pipeline_parallel/p2p_communication.py (hunk 10 above)
Comment (jaredcasper) [reply megatron/core/pipeline_parallel/p2p_communication.py]: I think it would make more logical sense to make `total_stages` and `current_stage` properties of the communicator. Then in schedules.py have `num_warmup_microbatches = total_stages - current_stage - 1`. What do you think?

File: megatron/core/pipeline_parallel/schedules.py (hunk 9 above)
Comment (jaredcasper) [reply megatron/core/pipeline_parallel/schedules.py:2010]: I see, makes sense.

File: megatron/core/pipeline_parallel/p2p_communication.py (hunk 10 above)
Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/p2p_communication.py]: makes sense. I think this would be better as num warmup batches is a scheduling concern! made the change to have current stage and total stages a property of communicator. Thanks for the suggestion!

**Comment by yashaswikarnati** [comment]
/ok to test [92b65d1](https://github.com/NVIDIA/Megatron-LM/pull/3129/commits/92b65d1e6e87a9a7d3d85e5856c6ef7f77a2581e)

**Comment by yashaswikarnati** [comment]
/ok to test [78ee58c](https://github.com/NVIDIA/Megatron-LM/pull/3129/commits/78ee58cc8e38fe20ccfec936fadc9b9d14beb516)

**Review by jaredcasper — APPROVED** [review]
